cpp/src/arrow/ipc/writer.cc

```diff
@@ -1020,13 +1014,30 @@ class ARROW_EXPORT IpcFormatWriter : public RecordBatchWriter {
   Status WriteDictionaries(const RecordBatch& batch) {
     ARROW_ASSIGN_OR_RAISE(const auto dictionaries, CollectDictionaries(batch, mapper_));
 
+    // TODO: Check for delta dictionaries. Can we scan for deltas while computing
+    // the RecordBatch payload to save time?
+
     for (const auto& pair : dictionaries) {
       IpcPayload payload;
       int64_t dictionary_id = pair.first;
       const auto& dictionary = pair.second;
 
+      // If a dictionary with this id was already emitted, check if it was the same.
+      // Note we don't compare dictionaries by value because it may be costly.
+      auto* last_dictionary = &last_dictionaries_[dictionary_id];
+      {
+        const auto maybe_last_dict = last_dictionary->lock();
+        if (maybe_last_dict && maybe_last_dict->data() == dictionary->data()) {
+          // Same dictionary data by pointer => no need to emit it again
+          continue;
+        }
+      }
+
       RETURN_NOT_OK(GetDictionaryPayload(dictionary_id, dictionary, options_, &payload));
       RETURN_NOT_OK(payload_writer_->WritePayload(payload));
```

Comment:
This should only be allowable if we are writing a stream? Dictionary replacements aren't allowed for IPC files

Reply:
Hmm, right. I should check how the file writer deals with this.

Reply:
Done.